src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -241,6 +249,11 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
             if (serviceType.IsConstructedGenericType &&
                 serviceType.GetGenericTypeDefinition() == descriptor.ServiceType)
             {
+                if (_callSiteCache.TryGetValue(new ServiceCacheKey(serviceType, slot), out ServiceCallSite serviceCallSite))
```

Comment:
Not a fan of having caching code all over the place but also don't have a better idea.

Reply:
I was initially trying to consolidate as well but didn't find a good place for it.